src/odemis/driver/keysight.py

```diff
@@ -481,56 +481,93 @@ def _set_power(self, p: bool) -> bool:
-                self.setVoltageMin(c + 1, lim[0])
-                self.setVoltageMax(c + 1, lim[1])
+        return p
 
-            # Reset all the settings by "setting" the VAs to the current values. We need this for 3 reasons:
-            # * if the VAs were changed while the power was off, they couldn't be applied, so we
-            #   need to set them now.
-            # * maybe the user has changed them in the meantime. So we need to put back the expected values.
-            # * it actually might be the first time the device is powered on, so we need to set them all.
-            try:
-                self.dutyCycle.value = self.dutyCycle.value
-            except TrueFormError as ex:
-                logging.warning("Failed to set the duty cycle to %s after power on: %s", self.dutyCycle.value, ex)
-            try:
-                self.period.value = self.period.value  # also updates .delay, in a safe way
-            except TrueFormError as ex:
-                logging.warning("Failed to set the period to %s after power on: %s", self.period.value, ex)
-
-            # Ends by enabling the output
-            self.setOutput(self._channel, True)
-            for c in self._tracking.keys():
-                self.setOutput(c, True)
-
-            if not self.requestLock():
-                # It's not a big deal if we can't lock the device. At worse the user might
-                # be able to change the settings manually, which might be a little confusing.
-                logging.warning("Failed to lock the device front panel")
-        else:  # Off
-            # Either disable the output (if off_voltage is None), or set a DC voltage
-            # Note: when power is off, the sync is still sent.
-            for c, volt in enumerate(self._off_voltage):
-                if volt is None:
-                    self.setOutput(c + 1, False)
-                else:
-                    # Note: in DC mode, some settings cannot be changed (like the phase & duty cycle).
-                    # Also, tracking doesn't work in DC mode, so should be re-enabled when powered on.
-                    self.applyDC(c + 1, volt)
+    def _power_device_on(self):
+        """
+        Configure the device to "on" state
+        """
+        # Powered on means a square waveform
+        try:
+            self.setWaveform(self._channel, "SQU")
+        except TrueFormError as ex:
+            # It happens that when setting the waveform, some current settings have to be reset,
+            # which the keysight reports as an error, but it's not an issue, as we set all the
+            # settings again afterwards. A second error would be a sign of a real issue.
+            logging.warning("Failed to set the waveform to square, will retry: %s", ex)
+            self.setWaveform(self._channel, "SQU")
+
+        # If required, set the other channel to track the main channel
+        for c, t in self._tracking.items():
```

Comment:
To me this is a bit confusing. I get that items in _tracking should follow the main channel, but now the non-tracking channel will be set as tracking via `setTracking`. Furthermore, it seems that this module only works with two channels. That will be one main and one in tracking. That means that _tracking will always be a list of max one, so the for-loop seems redundant.

Reply:
Yes, I agree with both remark. `setTracking` is strangely defined, but that's how the device works, so we cannot do anything. The reason for the loop is that originally we decided to make the driver generic enough so that if we want to support more devices from the same manufacturer with more channels, it would be easy to support. That means we end up for now with a strange loop that only loops 0 or once. In retrospect, it's probably over-engineering, but let's keep it like this (and this PR doesn't actually change this code, just move it around).